mesonbuild/modules/fs.py

```diff
@@ -310,6 +312,15 @@ def copyfile(self, state: ModuleState, args: T.Tuple[FileOrString, T.Optional[st
 
         return ModuleReturnValue(ct, [ct])
 
+    @noKwargs
+    @FeatureNew('fs.join_absolute', '1.1.0')
+    @typed_pos_args('fs.join_absolute', (str, File), (str, File))
+    def join_absolute(self, state: 'ModuleState', args: T.Tuple['FileOrString', 'FileOrString'], kwargs: T.Dict[str, T.Any]) -> str:
+        d1, d2 = str(args[0]), str(args[1])
+        if not PurePath(d2).is_absolute():
```

Comment:
On windows prefix does not have to be an absolute path, so this restriction might be pretty painful. Particularly because the fs module does not have a way of making a path absolute.

Reply:
Is that true? The way destdir_join() is implemented, it will not work if d2 is not absolute.

Reply:
@anarazel pretty sure prefix must be absolute on all platforms: https://github.com/mesonbuild/meson/blob/master/mesonbuild/coredata.py#L551

Reply:
prefix=/path/to/something works on windows, but is not an absolute path (it's a drive relative path).

Reply:
destdir_join historically only gets used internally when we can guarantee that all use cases only pass a value that's accepted by `meson setup --prefix=...`

How it's supposed to operate in other contexts is totally undefined.

Reply:
What I mean is that --prefix=/foo/bar works on windows, and isn't an absolute path.  To showcase I just made a throway branch of postgres, changing windows CI to set prefix to /frak:
https://github.com/anarazel/postgres/commit/c8a4f0aced0e98a14b2b0f86d22d622612d43677
CI shows it as being accepted:
https://cirrus-ci.com/task/6604479626215424?logs=configure#L321

I'm fairly certain there's a comment about that somewhere, I just don't recall where.

Reply:
I think that's an unexpected difference between is_absolute() and isabs():
```
>>> os.path.isabs('/foo')
True
>>> PurePath('/foo').is_absolute()
False
```

Interestingly, destdir_join() probably works in that case:
```
>>> PurePath('/foo/bar').parts[1:]
('foo', 'bar')
```

I guess the fix is to simply use `os.path.isabs()` here, I'll update.

Reply:
Fixed.

Reply:
> I think that's an unexpected difference between is_absolute() and isabs():

No, it's very expected. The general rule of pathlib is that you should ***expect*** it to do weird, insane things when you use it as object-oriented bindings for os.path functions.

What pathlib is good at is being a stdlib implementation of the os.Pathlike protocol. Every time we use it for *anything* else other than constructing a path object *or* using .as_posix(), it somehow turns out to be buggy...